### PR TITLE
refactor: add universe to restrictable enum sym

### PR DESCRIPTION
**Pull request by mlutze**
related to #5229

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 1]
```diff
@@ -413,7 +414,14 @@ object Symbol {
...
     * Restrictable Enum Symbol.
     */
-  final class RestrictableEnumSym(val namespace: List[String], val name: String, val loc: SourceLocation) extends Symbol {
+  final class RestrictableEnumSym(val namespace: List[String], val name: String, cases: List[Name.Ident], val loc: SourceLocation) extends Symbol {
+
+    // NB: it is critical that this be either a lazy val or a def, since otherwise `this` is not fully instantiated
+    /**
+      * The universe of cases associated with this restrictable enum.
+      */
+    lazy val universe: SortedSet[Symbol.RestrictableCaseSym] = cases.map(Symbol.mkRestrictableCaseSym(this, _)).to(SortedSet)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:423]: I my prefer a def.

File: main/src/ca/uwaterloo/flix/language/phase/unification/SetFormula.scala [hunk 2]
```diff
@@ -207,9 +207,10 @@ object SetFormula {
   /**
     * Creates an environment for mapping between proper types and formulas.
     */
-  def mkEnv(ts: List[Type], univ: List[Symbol.RestrictableCaseSym]): (Bimap[VarOrCase, Int], Set[Int]) = {
+  def mkEnv(ts: List[Type], univ: SortedSet[Symbol.RestrictableCaseSym]): (Bimap[VarOrCase, Int], Set[Int]) = {
     val vars = ts.flatMap(_.typeVars).map(_.sym).distinct.map(VarOrCase.Var)
-    val cases = (univ ++ ts.flatMap(_.cases)).distinct.map(VarOrCase.Case)
+    val cases = (univ.toSet ++ ts.flatMap(_.cases)).map(VarOrCase.Case)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/SetFormula.scala:212]: What's this? Looks strange indeed.

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 3]
```diff
@@ -413,7 +414,14 @@ object Symbol {
   /**
     * Restrictable Enum Symbol.
     */
-  final class RestrictableEnumSym(val namespace: List[String], val name: String, val loc: SourceLocation) extends Symbol {
+  final class RestrictableEnumSym(val namespace: List[String], val name: String, cases: List[Name.Ident], val loc: SourceLocation) extends Symbol {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:417]: I think we want to override equality... No?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:417]: Ah, we already have the right equality, correct?

**Comment by magnus-madsen** [comment]
Looks good, but I added some comments.

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:417]: Correct

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:423]: What is a lazy val, if not a cached def?

I guess this wouldn't exactly be a bottleneck.

**Pull request merged**